async-nats/src/jetstream/pull.rs

```diff
@@ -0,0 +1,192 @@
+        .await
+    }
+
+    pub async fn process(&mut self, batch: usize) -> PullStream<'_> {
+        PullStream::new(self, batch)
+    }
+}
+
+pub struct BatchStream<'a> {
+    pending_messages: usize,
+    subscriber: &'a mut Subscriber,
+    context: Context,
+}
+
+impl<'a> BatchStream<'a> {
+    async fn batch(
+        batch: BatchOptions,
+        consumer: &'a mut Consumer<PullConsumerConfig>,
+    ) -> Result<BatchStream<'a>, Error> {
+        if consumer.subscription.is_none() {
+            consumer.subscription = Some(
+                consumer
+                    .context
+                    .client
+                    .subscribe(consumer.inbox.clone())
+                    .await?,
+            );
+        }
+        consumer.request_batch(batch).await?;
+
+        Ok(BatchStream {
+            pending_messages: batch.batch,
+            subscriber: consumer.subscription.as_mut().unwrap(),
+            context: consumer.context.clone(),
+        })
+    }
+}
+
+impl<'a> Stream for BatchStream<'a> {
+    type Item = Result<JetStreamMessage, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        if self.pending_messages == 0 {
+            return std::task::Poll::Ready(None);
+        }
+        match self.subscriber.receiver.poll_recv(cx) {
+            std::task::Poll::Ready(message) => {
+                if let Some(message) = message {
+                    if message.status == Some(StatusCode::TIMEOUT) {
+                        return Poll::Ready(None);
+                    }
+                    if message.status == Some(StatusCode::IDLE_HEARBEAT) {
+                        return std::task::Poll::Pending;
+                    }
+                    self.pending_messages -= 1;
+                    Poll::Ready(Some(Ok(JetStreamMessage {
+                        context: self.context.clone(),
+                        message,
+                    })))
+                } else {
+                    Poll::Ready(None)
+                }
+            }
+            std::task::Poll::Pending => std::task::Poll::Pending,
+        }
+    }
+}
+
+pub struct PullStream<'a> {
+    consumer: &'a mut Consumer<PullConsumerConfig>,
+    batch: usize,
+}
+
+impl<'a> PullStream<'a> {
+    fn new(consumer: &'a mut Consumer<PullConsumerConfig>, batch: usize) -> PullStream<'a> {
+        PullStream { consumer, batch }
+    }
+
+    pub fn process(&'a mut self) -> impl Stream<Item = Result<JetStreamMessage, Error>> + 'a {
+        try_stream! {
+            loop {
+                let  stream = self.consumer.batch(self.batch, None).await?;
+                for await message in stream {
+                    yield message?;
+                }
+            }
+        }
+    }
+}
+
+/// Used for next Pull Request for Pull Consumer
+#[derive(Debug, Default, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+pub struct BatchOptions {
```

Comment:
Not an options builder, we've been pretty consistent this time with options vs configs.

```suggestion
pub struct BatchConfig {
```